packages/core/src/compiler/Style.ts

```diff
@@ -1640,12 +1662,195 @@ const translateSubstsBlock = (
+    return emptyErrs();
+  } else if (expr.tag === "GPIDecl") {
+    const e1: StyleResults = checkGPIInfo(selEnv, expr);
+    const e2: StyleResults[] = expr.properties.map((p) => check(p.value));
+    return flatErrs([e1].concat(e2));
+  } else if (expr.tag === "Layering") {
+    return flatErrs([check(expr.below), check(expr.above)]);
+  } else if (expr.tag === "PluginAccess") {
+    return flatErrs([check(expr.contents[1]), check(expr.contents[2])]);
+  } else if (expr.tag === "Tuple") {
+    return flatErrs([check(expr.contents[0]), check(expr.contents[1])]);
+  } else if (expr.tag === "VectorAccess") {
+    return check(expr.contents[1]);
+  } else if (expr.tag === "MatrixAccess") {
+    return flatErrs(expr.contents[1].map(check));
+  } else if (
+    expr.tag === "Fix" ||
+    expr.tag === "Vary" ||
+    expr.tag === "StringLit" ||
+    expr.tag === "BoolLit"
+  ) {
+    return emptyErrs();
+  } else {
+    console.error("expr", expr);
+    throw Error("unknown tag");
+  }
+};
+
+const checkBlockPath = (selEnv: SelEnv, path: Path): StyleResults => {
+  // TODO(errors) / Block statics
+  // Currently there is nothing to check for paths
+  return emptyErrs();
+};
+
+const checkLine = (
+  selEnv: SelEnv,
+  line: Stmt,
+  acc: StyleResults
+): StyleResults => {
+  if (line.tag === "PathAssign") {
+    const pErrs = checkBlockPath(selEnv, line.path);
+    const eErrs = checkBlockExpr(selEnv, line.value);
+    return combineErrs(combineErrs(acc, pErrs), eErrs);
+  } else if (line.tag === "Override") {
+    const pErrs = checkBlockPath(selEnv, line.path);
+    const eErrs = checkBlockExpr(selEnv, line.value);
+    return combineErrs(combineErrs(acc, pErrs), eErrs);
+  } else if (line.tag === "Delete") {
+    const pErrs = checkBlockPath(selEnv, line.contents);
+    return combineErrs(acc, pErrs);
+  } else {
+    throw Error(
+      "Case should not be reached (anonymous statement should be substituted for a local one in `nameAnonStatements`)"
+    );
+  }
+};
+
 const checkBlock = (selEnv: SelEnv, block: Block): StyleErrors => {
-  // TODO: Block checking and return block errors, not generic sty errors
-  // Static semantics would go here
-  return [];
+  // Block checking; static semantics
+  // The below properties are checked in one pass (a fold) over the Style AST:
+
+  // Check that every shape name and shape property name in a shape constructor exists
+  // Check that every function, objective, and constraint exists
+  // NOT CHECKED as this requires more advanced env-building work: At path construction time, check that every Substance object exists in the environment of the block + selector, or that it's defined as a local variable
+
+  const res: StyleResults = block.statements.reduce(
+    (acc: StyleResults, stmt: Stmt): StyleResults =>
+      checkLine(selEnv, stmt, acc),
+    emptyErrs()
+  );
+
+  // TODO(errors): Return warnings (non-fatally); currently there are no warnings though
+  if (res.warnings.length > 0) {
```

Comment:
Looking at the comment: should this one throw, or you are supposed to do something nonfatal here?

Reply:
Yeah, it should throw as a reminder to deal better with warnings in the future. There are currently no warnings in the system, so this is fine. But we should think about how to aggregate/accumulate them in the future, as currently there are only errors, which are reported as if everything is fatal.